Add Visvalingam simplification to cleanfigure

widdma opened the pull request:
This patch replaces the `minimumPointsDistance` feature with a `targetResolution` option. The cleanfigure script then performs path simplification with Visvalingam's algorithm using a tolerance based on `targetResolution`. 

Essentially, cleanfigure now removes points which do not add any detail at the specified target resolution.

PeterPablo commented:
See issue #281.

@widdma Thank you for your contribution! Do you have a minimal example that nicely shows the performance of the new algorithm? We might want to either adapt an existing ACID test or introduce a new one.

egeerardyn commented on src/cleanfigure.m:
```diff
@@ -50,15 +50,15 @@ function cleanfigure(varargin)
   m2t.cmdOpts = m2tInputParser;
-  m2t.cmdOpts = m2t.cmdOpts.addParamValue(m2t.cmdOpts, 'minimumPointsDistance', 1.0e-10, @isnumeric);
   m2t.cmdOpts = m2t.cmdOpts.addParamValue(m2t.cmdOpts, 'handle', gcf, @ishandle);
+  m2t.cmdOpts = m2t.cmdOpts.addParamValue(m2t.cmdOpts, 'targetResolution', '15x9@600',@isstr);
```
It is advised to use `@ischar` for checking a string. 

Anyhow, I think we're better off with a more structured format that we can validate more easily. I'd propose something like `[15 9 600]` or `{15, 9, 600}`. Those we can check, and we don't need to parse them at the time we use them.

We also have to provide some backward compatibility: the `minimumPointsDistance` parameter should remain valid for at least 1 release (even if it doesn't do anything). So I would propose to add:

```
m2t.cmdOpts = m2t.cmdOpts.addParamValue(m2t.cmdOpts, 'minimumPointsDistance', 1.0e-10, @isnumeric);

m2t.cmdOpts = m2t.cmdOpts.deprecateParam(m2t.cmdOpts, 'minimumPointsDistance', 'targetResolution')
```

okomarov replied:
We can try to translate the 600 parameter into the minimum point distance and viceversa (when the user insists on using `'minimumPointsDistance'`), hence keeping backward compatibility in the syntax. 

Also, we could allow for double syntax `targetResolution` to make it easier to play with the algo:
* scalar, that sets only the points per area
* all 3 parameters in a vector, for more advanced tweaking.

For the rest I agree with @egeerardyn.

widdma replied:
Because we (currently) only use the total number of points/pixels in the target image, ie. product of these 3 parameters, perhaps we should just use that as the parameter and rename it to something like `targetNumberOfPoints`? We can then have a note on calculating this parameter in the description .

Alternatively, we could expose the bare `tol` parameter and allow the user to do their own `figArea/targetNumberOfPoints` calculation. This gives the user more freedom but means a good default is not possible.

Maybe we could expose both:
  -  A `minimumFeatureArea` option for fine grained control to set manually the `tol` parameter of Visvalingam algorithm,
  -  A `targetResolution` option which selects a good `tol` for `minimumFeatureArea` based on the figure properties.

If we want to keep support for `minimumPointsDistance` with this function, having `tol=sqrt(2*minimumPointsDistance)` seems like a good choice, but of course won't be identical.

egeerardyn replied:
@widdma : Fine by me. As long as the parameters we expose are somewhat orthogonal to each other and understandable to our users, it's fine by me.

PeterPablo commented:
Please furthermore adjust the ACID test, that makes use of `minimumPointsDistance`: `plain_cos()`.

egeerardyn commented on src/cleanfigure.m:
```diff
@@ -6,8 +6,8 @@ function cleanfigure(varargin)
-%   CLEANFIGURE('minimumPointsDistance',DOUBLE,...) explicitly specified the
-%   minimum distance between two points. (default: 1.0e-10)
+%   CLEANFIGURE('targetResolution','WxH@Res',...) explicitly specify the
+%   target resolution of the path simplification. (default: '15x9@600')
```
I think we need a bit more explanation on this parameter to make sense of the three numbers that are given. I think it also should be mentioned that the dimensions of the 2 first ones are "length" (e.g. centimeter or inch) and the last is "points per [same unit]".

egeerardyn commented on src/cleanfigure.m:
```diff
@@ -367,6 +368,259 @@ function coarsenLine(meta, handle, minimumPointsDistance)
+  if numel(xData) <= 2
+      return;
+  end
+  if strcmpi(targetResolution,'off')
```
Wouldn't it make more sense to have the algorithm switched off if the resolution (`param(3)`) is infinity or NaN?

egeerardyn commented on src/cleanfigure.m:
```diff
@@ -367,6 +368,259 @@ function coarsenLine(meta, handle, minimumPointsDistance)
+    while child > 1
+      parent = bitshift(n,-1);
+      if a(heap(child)) < a(heap(parent)) ...
+              || (a(heap(child)) == a(heap(parent)) && rand(1)>0.5)
```
What's that `rand` doing here?
Is it really essential that the algorithm is randomized? If not: please hard-code a value, since such random behavior makes testing much harder.

widdma replied:
rand removed.

egeerardyn commented on src/cleanfigure.m:
```diff
@@ -367,6 +368,259 @@ function coarsenLine(meta, handle, minimumPointsDistance)
+      % Find the minimum of the root and its children
+      swap = root;
+      if a(heap(lchild)) < a(heap(swap)) ...
+          || (a(heap(lchild)) == a(heap(swap)) && rand(1)>0.5)
```
Same remark as https://github.com/matlab2tikz/matlab2tikz/pull/571/files#r26300183: `rand` is bad for testing.

egeerardyn commented on src/cleanfigure.m:
```diff
@@ -367,6 +368,259 @@ function coarsenLine(meta, handle, minimumPointsDistance)
+        swap = lchild;
+      end
+      if rchild <= len && (a(heap(rchild)) < a(heap(swap)) ...
+          || (a(heap(rchild)) == a(heap(swap)) && rand(1)>0.5))
```
Same remark as https://github.com/matlab2tikz/matlab2tikz/pull/571/files#r26300183: `rand` is bad for testing.

egeerardyn commented on src/cleanfigure.m:
```diff
@@ -367,6 +368,259 @@ function coarsenLine(meta, handle, minimumPointsDistance)
+        pos(heap(swap)-1) = root;
+
+        %swap in the heap array
+        t = heap(root);
```
You can swap 2 variables `a` and `b` more easily using `[a, b] = deal(b, a);`.
Inside an array, it's even easier to do the following:
```matlab
heap([root swap]) = heap([swap root])
```
for `root` and `swap` of identical dimensions.

egeerardyn commented:
@widdma This contribution is very welcome and I really like most of it, so most of my comments are minor things on style.

However, two significant points that we should look at are:

* [ ] the format of the input parameter (and some backwards compatibility for the old parameter)
* [ ] the use of `rand` in the algorithm

okomarov commented on src/cleanfigure.m:
```diff
@@ -367,6 +368,259 @@ function coarsenLine(meta, handle, minimumPointsDistance)
+  if ~isempty(zData)
+    % Don't simplify 3d plots
+    return;
+  end
```
The algorithm does not work in 3d?

PeterPablo replied:
Quick searching yields: http://www.sciencedirect.com/science/article/pii/S0097849398000375
It seems it is possible.

okomarov replied:
We could probably work with a projection of z, however let's put a todo note there.

egeerardyn replied:
Regarding 3D: let's indeed put a TODO. 
Probably one of the reasons why this hasn't been addressed (in our old code and here), is because in 3D there are more intricate things that can be done improperly than in 2D:
- we need to reduce the number of points (e.g. with such an algorithm)
- the matter of perspective / view angle comes into play to determine whether a feature is distinguishable or not
- the data formats used by the different objects is more complicated (e.g. X/Y vectors or possibly meshgridded matrices and Z a matrix)
- the simplified data matrix needs to comply to the same constraints: I expect that we can only remove a complete row or column of the Z matrix in general, except maybe for line plots (which is a direct generalization of this code, but with 3D points).

widdma replied:
While the algorithm does indeed extend to 3D quite nicely, along with the issues @egeerardyn pointed out, selecting a good tolerance level is not quite as easy as the 2D case.

It's certainly a TODO.

okomarov commented on src/cleanfigure.m:
```diff
@@ -367,6 +368,259 @@ function coarsenLine(meta, handle, minimumPointsDistance)
+  isXlog = strcmp(get(meta.gca, 'XScale'), 'log');
+  vxData = xData;
+  if isXlog
+	  vxData = log10(xData);
```
One indentation only here

widdma replied:
Thanks, by switching between the Matlab editor and vim, I seem to have mixed tabs and spaces (the horror!). I'll fix my indentation up. Is there a prefered shift width for this project?

egeerardyn replied:
4 spaces per indent is preferred, but to be honest we still have quite some chunks with 2 spaces per indentation level. In any case: 8 is too much :wink:

widdma replied:
Great, I find 4 spaces much easier to read in matlab code, but most of the surrounding stuff was 2 spaces, so I went with that. 

Converted my sections to 4 spaces indentation.

okomarov commented on src/cleanfigure.m:
```diff
@@ -367,6 +368,259 @@ function coarsenLine(meta, handle, minimumPointsDistance)
+
+  %Split up lines which are seperated by NaNs
+  nans = isnan(vxData) | isnan(vyData);
+  idx = sort([1,find(nans)-1,find(nans)+1,numel(vxData)]);
```
Would this fail on consecutive nans, e.g. [1,2,3,nan,nan,10,20]?

egeerardyn replied:
@okomarov I think that indeed this code assumes there are no consecutive NaNs.

widdma replied:
@okomarov Thanks, I've added your snippet. Unless anyone has any objections, I'm replacing consecutive of NaNs with just one NaN.

okomarov commented on src/cleanfigure.m:
```diff
@@ -367,6 +368,259 @@ function coarsenLine(meta, handle, minimumPointsDistance)
+  set(handle, 'XData', xData);
+  set(handle, 'YData', yData);
+
+  return;
```
Unnecessary(?)

widdma replied:
Thanks, now removed.

okomarov commented:
Some general remarks about the algo. 
It is not Matlabish style, and it is fine as long as it works well :smile:, in terms of simplication results on some test cases and performance. 

Maybe @widdma if you used some sources, could you reference them (if not already done) in the help part of the algo?

PeterPablo commented:
Searching GitHub I found several implementations in C and Python. [This one](https://github.com/code42day/vis-why/blob/master/index.js) comes with easy readable variable and function names. I posted a link to the original research paper in the above referenced issue.

egeerardyn commented:
@PeterPablo I also found a Python implementation by @widdma [here](https://github.com/widdma/tikz-scripts/blob/master/tikzsimplify).

@okomarov I agree the style is not really MATLAB-style, but I wholeheartedly agree that having (reasonably fast) and working code is already a great first step. If it really bothers us, we can always come back to the algorithm and rewrite it in MATLAB style (although that is not always straightforward starting from an algorithm description or C-ish code).

widdma commented:
Thank you all for the encouragement and helpful criticism. I'll address general criticisms here and particulars in the line notes. I'm somewhat new to Github, so please if this is not the best way to do things.

@okomarov and @egeerardyn, I agree that it's not particularly MATLAB-style code which was a deliberate choice to keep the code fast; however, it's still somewhat slower than I'd like. I've tried to document my use of implicit data structures, but if you consider this too difficult to read/maintain, I can rewrite it in a more MATLAB-style.

On the use of `rand`: it's not necessary and can safely removed. This was to prevent the simplification from preferentially removing points at the start of the path. This came up in cases with periodic functions and even grid spacing.

egeerardyn commented:
@widdma At one point or another, I think we should rewrite it to more MATLAB-ese, but for me the code style wouldn't stop me from merging this PR. If you address the other points we've raised, that'd be :+1: 

If you have any more questions (related to this or to git/github): just shoot :smile: . But it's indeed good practice to reply to specifics in the commits and the overall stuff in the PR itself. The specifics will fold once they are addressed, so this keeps the high-level overview lean.

widdma commented:
Should this be enabled by default? It is quite a large change to apply, but then again, they are calling `cleanfigure`.

egeerardyn commented:
@widdma If the results and performance are comparable (or better) than before, I'd vote to enable it by default.

PeterPablo commented:
Me too. I see a crucial step in determining a sensible default value, in case the user does not specify one.

okomarov commented:
Unfortuantely, performance is not brilliant.
I tested it on `ACID(47)` (the slow test) with the .fig provided by Egon at [mat files](https://dl.dropboxusercontent.com/u/452005/m2t/acid47/index.html).

It reduces points from ~5e4 to 148, which is nice, but it takes 14seconds on my i7-2600. I reckon we should try to improve `pop()`, `push()` and `down()`.

![untitled](https://cloud.githubusercontent.com/assets/3731173/6638597/9340e468-c97f-11e4-87cb-003e043348a1.png)

okomarov commented on src/cleanfigure.m:
```diff
@@ -298,73 +300,253 @@ function pruneOutsideBox(meta, handle)
+    pos = [0;(1:len)';0];
+
+    area = @(i,j,k) abs((y(j) - y(k)).*x(i) + (y(k)-y(i)).*x(j) + ...
+        (y(i)-y(j)).*x(k))/2;
```
Add comment: area by schoelace formula.

okomarov commented:
I tried to look into the algo (also with some help from http://en.wikipedia.org/wiki/Binary_heap) and it is indeed intuitive that it should run faster than calling a `min()` every time an area is updated. However, the fact that `min` is builtin could outpace the interpreted implementation in Matlab. 

One last attempt is to initialize the `pos`, `heap` as `uint*` and area as single. Should be faster, but might not be enough. 

@matlab2tikz/developers How about .mex (ing) `featureArea`?

widdma commented:
@okomarov yeah, it is certainly slower than I would like. While algorithm is inherently O(nlog(n)), the matlab interpreter seems to be giving it quite a large constant of proportionality.

It seems like a lot of time is lost in function call overhead. If I explicitly expand all the function calls, the runtime on my machine for 1e5 points, as measured by `tic;cleanfigure;toc`, drops from 13.4s to 6.5s. (I'm only have access to matlab 2012b right now, which has a lot of overhead when running profile. Runtimes are closer to 70s with profiling on)

Of course, this leads to some terribly unwieldy code. I've put it in the branch `line-simp-test`

Using mex is an option, but it introduces distribution problems.

egeerardyn commented:
@okomarov I would refrain from MEXing this. As @widdma indicates, this creates a lot of distribution problems: if we provide binaries, we need to provide it for a lot of OSes, MATLAB versions, ... If we don't we make our tool less usable for a lot of people. I know that in my department, most people don't have a compiler installed by default (let alone have it configured). I doubt anybody would install a compiler specifically to export some plots.

MEX is nice-to-have, but I'd prefer to have everything in MATLAB code. Having small parts also (!) available in MEX is something we might consider. But for me that's a last resort.

@widdma I've also observed that call overhead can be quite high in MATLAB.

okomarov commented:
I agree, and have not thought about the distro nightmare :|

okomarov commented:
In terms of performance, I reckon we can leave the code as is and check that `testHeadless()` completes in the ~same time as without this PR. 

Basically, for now I am considering the impact on **development** and I expect that the cleaning time will trade away compile time in our favor, see ACID(47). (This won't be true on Win, since raw ACID(47) takes 4s to compile and 12s to clean, but I am the minority here and do not have the Tollbox anyways :smile:)

For the **user**, I consider the issue from two perspectives:
* the current algo is an improvement on the previous clenfigure which e.g. has no effect on ACID(47). Thus, reduces memory problems in latex, size of figures and compile time. A few seconds of wait time are a huge improvement on minutes on trying to figure out how to avoid memory issues in Latex. 
* I tend to think that plots with lots of points are somewhat an edge case (I might be wrong here) and that the average user never uses cleanfigure. 

I also think that this algo requires more resources but operates in a smarter way. Hence, to improve the overall cleanfigure performance we could first call a method that makes sure that no more than a data point falls into a pixel (up to some zoom multiplier) and then release Visvalingam for the precision cut. 

For me, to complete the PR we need to address the input case raised by Egon (and add a few additional notes on the algo with some links, which we can do)

widdma commented:
In order to address the speed concerns, I've changed `simplifyLine` to discretize  the data to a quarter of a pixel width before performing full path simplification. While this isn't going to be performing true Visvalingam simplification any more, it's certainly a good approximation with the chosen `tol`. 

Simplification on the fig file provided by @okomarov from ACID(47) now performs in under 1s on my current setup (Matlab 2014a, i5-2500K). Simplifying 5e6 points via 
```matlab
x  = linspace(0,2*pi,5e6);
y = sin(x);
plot(x,y);
cleanfigure;
```
takes about 10s. But the bottleneck now appears to be `segmentVisible` rather than the simplification code. It may be worth optimizing this in the future. With this level of performance, I would be much more comfortable enabling this feature by default.

I've also addressed the issue @egeerardyn raised with the `targetResolution` parameter. This is now an array in the format `[W,H,Res]`. With the recent changes to `simplifyLine`, all 3 parameters are now explicitly needed.

okomarov commented:
Great job, the sine plot with 5e6 cleans in less than a second on my i7-2600!
I am currently reviewing a few minor points and have some questions.

The `[W, H, res]` specification is somewhat elusive to me. 
I guess if the user is (physically) printing the figure, she must have a pretty good idea of the size and the dpi and assuming that the monitor has same resolution you can think of res in terms of PPI, i.e. `[W, H, PPI]` (assuming the unit is inches)?
Then, I think in this context `W` and `H` are somewhat superfluous, in the sense they can be queried from the gca. 
In that context, the zoom multiplier I mentioned in another comment makes sure that full monitor visualization is still at the PPI resolution (and if the user knows what he wants and calls [W,H,res], then you do not need the zoom multiplier). 

I am raising this point because I was thinking to *eventually* allow double syntax on `targetResolution` (in another PR). The base case would be `'targetResolution', PPI` where ppi is scalar and the current 3 element input.

egeerardyn commented on src/cleanfigure.m:
```diff
@@ -50,15 +56,18 @@ function cleanfigure(varargin)
   m2t.cmdOpts = m2tInputParser;
-  m2t.cmdOpts = m2t.cmdOpts.addParamValue(m2t.cmdOpts, 'minimumPointsDistance', 1.0e-10, @isnumeric);
   m2t.cmdOpts = m2t.cmdOpts.addParamValue(m2t.cmdOpts, 'handle', gcf, @ishandle);
+  m2t.cmdOpts = m2t.cmdOpts.addParamValue(m2t.cmdOpts, 'targetResolution', [15 9 600], @(a)numel(a) == 3);
```
Please just write a proper function to validate this.
```matlab
function bool = isValidTargetResolution(val)
    bool = isnumeric(val) && numel(v) == 3;
```
I haven't look at the rest how `NaN` are handled, but if those cause failures, they should be caught at this location as well.

PeterPablo commented:
Yesterday and today I used `m2t` with this PR and I am very happy with the results and did not run into issues.

okomarov commented:
I reviewed the PR and added a bugfix and some help/comment edits. Issued a new PR #585 which we can build on if needed (it's on upstream branch too). Closing the current one.
